File: shared/contracts/tests/LiveWireRoundTrip.swift
import Foundation

@main enum LiveWireRoundTrip {
    static func roundTrip<T: Codable>(_ type: T.Type, _ data: Data) throws -> Data {
        try JSONEncoder().encode(JSONDecoder().decode(type, from: data))
    }
    static func main() throws {
        let rows = try JSONSerialization.jsonObject(with: Data(contentsOf: URL(fileURLWithPath: CommandLine.arguments[1]))) as! [[String: Any]]
        for row in rows {
            let data = try JSONSerialization.data(withJSONObject: row["value"]!)
            let valid = row["valid"] as! Bool
            do {
                let encoded: Data
                switch row["schema"] as! String {
                case "LiveTransport": encoded = try roundTrip(LiveTransportDTO.self, data)
                case "LiveSession": encoded = try roundTrip(LiveSessionDTO.self, data)
                case "LiveSessionStatus": encoded = try roundTrip(LiveSessionStatusDTO.self, data)
                case "CurrentLiveSession": encoded = try roundTrip(CurrentLiveSessionDTO.self, data)
                case "LiveSessionCreateRequest": encoded = try roundTrip(LiveSessionCreateRequestDTO.self, data)
                case "HostedHelperStreamEvent": encoded = try roundTrip(HostedHelperStreamEventDTO.self, data)
                case "LiveCapabilities": encoded = try roundTrip(LiveCapabilitiesDTO.self, data)
                case "HostedHelperUsage": encoded = try roundTrip(HostedHelperUsageDTO.self, data)
                case "HostedHelperResult": encoded = try roundTrip(HostedHelperResultDTO.self, data)
                case "ErrorResponse": encoded = try roundTrip(ErrorResponseDTO.self, data)
                case "LiveHistoryMessage": encoded = try roundTrip(LiveHistoryMessageDTO.self, data)
                default: fatalError("Unknown fixture")
                }
                precondition(valid, "Invalid fixture accepted")
                let before = try JSONSerialization.jsonObject(with: data) as! NSDictionary
                let after = try JSONSerialization.jsonObject(with: encoded) as! NSDictionary
                precondition(before == after, "Wire shape changed (missing/null)")
            } catch is DecodingError { precondition(!valid, "Valid fixture rejected") }
        }
        print("Swift shared wire fixtures: \(rows.count) PASS")
    }
}

File: shared/contracts/tests/LiveWireRoundTrip.kt
import chat.mural.contracts.*
import kotlinx.serialization.*
import kotlinx.serialization.json.*
import java.io.File

private inline fun <reified T> roundTrip(value: JsonElement): JsonElement =
    Json.encodeToJsonElement(Json.decodeFromJsonElement<T>(value))

fun main(args: Array<String>) {
    val rows = Json.parseToJsonElement(File(args[0]).readText()).jsonArray
    for (row in rows) {
        val entry = row.jsonObject
        val value = entry.getValue("value")
        val valid = entry.getValue("valid").jsonPrimitive.boolean
        try {
            val encoded = when (entry.getValue("schema").jsonPrimitive.content) {
                "LiveTransport" -> roundTrip<LiveTransportDTO>(value)
                "LiveSession" -> roundTrip<LiveSessionDTO>(value)
                "LiveSessionStatus" -> roundTrip<LiveSessionStatusDTO>(value)
                "CurrentLiveSession" -> roundTrip<CurrentLiveSessionDTO>(value)
                "LiveSessionCreateRequest" -> roundTrip<LiveSessionCreateRequestDTO>(value)
                "HostedHelperStreamEvent" -> roundTrip<HostedHelperStreamEventDTO>(value)
                "LiveCapabilities" -> roundTrip<LiveCapabilitiesDTO>(value)
                "HostedHelperUsage" -> roundTrip<HostedHelperUsageDTO>(value)
                "HostedHelperResult" -> roundTrip<HostedHelperResultDTO>(value)
                "ErrorResponse" -> roundTrip<ErrorResponseDTO>(value)
                "LiveHistoryMessage" -> roundTrip<LiveHistoryMessageDTO>(value)
                else -> error("Unknown fixture")
            }
            check(valid) { "Invalid fixture accepted" }
            check(encoded == value) { "Wire shape changed (missing/null)" }
        } catch (error: SerializationException) { check(!valid) { "Valid fixture rejected: $error" } }
    }
    println("Kotlin shared wire fixtures: ${rows.size} PASS")
}
